### ISSUE TITLE
(update: provider args conflicts need better description) Error: failed to create SAML application / group: invalid private key

### ISSUE THREAD

**Issue by JPLachance** [okta/terraform-provider-okta#983]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
Terraform v1.0.2
on linux_amd64
+ provider registry.terraform.io/hashicorp/aws v3.63.11
+ provider registry.terraform.io/okta/okta v3.21.0
```

### Can we manage Groups and Apps then using a Client ID / Private key combo?

Greetings,

I'm trying to manage Groups and Apps using an Okta Application of type API Services:
> Interact with Okta APIs using the scoped OAuth 2.0 access tokens for machine-to-machine authentication.

So, I created an APP, granted it access to:
- okta.apps.manage
- okta.apps.read
- okta.groups.manage
- okta.groups.read
- okta.roles.manage
- okta.roles.read

Then, I tried the following Terraform code. The `terraform plan` works, the `terraform apply` fails:
```log
okta_group.example: Creating...
okta_app_saml.example: Creating...
╷
│ Error: failed to create group: invalid private key
│ 
│   with okta_group.example,
│   on test.tf line 25, in resource "okta_group" "example":
│   25: resource "okta_group" "example" {
│ 
╵
╷
│ Error: failed to create SAML application: invalid private key
│ 
│   with okta_app_saml.example,
│   on test.tf line 30, in resource "okta_app_saml" "example":
│   30: resource "okta_app_saml" "example" {
│ 
╵
```

Since a `terraform plan` works, I suspect my private key is valid.

Therefore, is it possible to use a `client_id` with a `private_key` to manage Okta Apps and Groups?

### Affected Resource(s)

- okta_app_saml
- okta_group

### Terraform Configuration Files

```hcl
terraform {
  required_providers {
    okta = {
      source  = "okta/okta"
      version = "3.21.0"
    }
  }
}

provider "okta" {
  org_name    = "coveo-cloud"
  base_url    = "okta.com"
  client_id   = "Some"
  private_key = "Secret"
  scopes      = [
    "okta.apps.manage",
    "okta.apps.read",
    "okta.groups.manage",
    "okta.groups.read",
    "okta.roles.manage",
    "okta.roles.read",
  ]
}

resource "okta_group" "example" {
  name        = "Example"
  description = "My Example Group"
}

resource "okta_app_saml" "example" {
  label                    = "example"
  sso_url                  = "https://example.com"
  recipient                = "https://example.com"
  destination              = "https://example.com"
  audience                 = "https://example.com/audience"
  subject_name_id_template = "$${user.userName}"
  subject_name_id_format   = "urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress"
  response_signed          = true
  signature_algorithm      = "RSA_SHA256"
  digest_algorithm         = "SHA256"
  honor_force_authn        = false
  authn_context_class_ref  = "urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport"

  attribute_statements {
    type         = "GROUP"
    name         = "groups"
    filter_type  = "REGEX"
    filter_value = ".*"
  }
}
```

### Expected Behavior

I would expect the `terraform apply` to work.

### Actual Behavior

It does not.

### Steps to Reproduce

1. `terraform apply`

### Notes

I suspect I need a Service account with Admin access and an API key for the provider to work as expected, but I cannot find a clear statement about when to use a `client_id` with a `private_key` VS when to use an `api_token` in the Okta Terraform provider documentation. From what I understand, my issue is purely around missing documentation.

Addendum from @monde . If the TF and env vars are set correctly then there isn't an SAML/API error. As @JPLachance points out, the documentation is not clear here.

**Comment by monde** [comment]
I will look into this @JPLachance

**Comment by monde** [comment]
@JPLachance I think your tf settings are messing this up for you. This is what I get:

my main.tf:

```terraform
terraform {
  required_providers {
    okta = {
      source = "okta/okta"
    }
  }
}

resource "okta_group" "example" {
  name        = "Example"
  description = "My Example Group"
}

resource "okta_app_saml" "example" {
  label                    = "example"
  sso_url                  = "https://example.com"
  recipient                = "https://example.com"
  destination              = "https://example.com"
  audience                 = "https://example.com/audience"
  subject_name_id_template = "$${user.userName}"
  subject_name_id_format   = "urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress"
  response_signed          = true
  signature_algorithm      = "RSA_SHA256"
  digest_algorithm         = "SHA256"
  honor_force_authn        = false
  authn_context_class_ref  = "urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport"

  attribute_statements {
    type         = "GROUP"
    name         = "groups"
    filter_type  = "REGEX"
    filter_value = ".*"
  }
}
```

```shell
export OKTA_API_TOKEN=XXXX
export OKTA_ORG_NAME=myorg
export OKTA_BASE_URL=okta.com

terraform init
terraform apply

# ....
Terraform will perform the following actions:

  # okta_app_saml.example will be created
  + resource "okta_app_saml" "example" {
      + accessibility_self_service = false
      + audience                   = "https://example.com/audience"
      + authn_context_class_ref    = "urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport"
      + auto_submit_toolbar        = false
      + certificate                = (known after apply)
      + destination                = "https://example.com"
      + digest_algorithm           = "SHA256"
      + entity_key                 = (known after apply)
      + entity_url                 = (known after apply)
      + hide_ios                   = false
      + hide_web                   = false
      + honor_force_authn          = false
      + http_post_binding          = (known after apply)
      + http_redirect_binding      = (known after apply)
      + id                         = (known after apply)
      + key_id                     = (known after apply)
      + label                      = "example"
      + logo_url                   = (known after apply)
      + metadata                   = (known after apply)
      + metadata_url               = (known after apply)
      + name                       = (known after apply)
      + recipient                  = "https://example.com"
      + response_signed            = true
      + saml_version               = "2.0"
      + sign_on_mode               = (known after apply)
      + signature_algorithm        = "RSA_SHA256"
      + skip_groups                = false
      + skip_users                 = false
      + sso_url                    = "https://example.com"
      + status                     = "ACTIVE"
      + subject_name_id_format     = "urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress"
      + subject_name_id_template   = "${user.userName}"
      + user_name_template         = "${source.login}"
      + user_name_template_type    = "BUILT_IN"

      + attribute_statements {
          + filter_type  = "REGEX"
          + filter_value = ".*"
          + name         = "groups"
          + namespace    = "urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified"
          + type         = "GROUP"
        }
    }

  # okta_group.example will be created
  + resource "okta_group" "example" {
      + description = "My Example Group"
      + id          = (known after apply)
      + name        = "Example"
      + skip_users  = false
    }

Plan: 2 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

okta_group.example: Creating...
okta_app_saml.example: Creating...
okta_group.example: Creation complete after 0s [id=00g15ptyfqkPEKunI0h8]
okta_app_saml.example: Creation complete after 1s [id=0oa15ptz2j9msjJFW0h8]

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
```

**Comment by monde** [comment]
@JPLachance I think filling out the `provider "okta" {}` block values tripped up the provider set up. There may be a bug there (e.g. the provider should be smarter about config it seems), but if you do the simplest thing for config using ENV vars then you can get this to work. I wanted to show you that in case you were stuck.

I'm going to keep this open for config bug value.

**Comment by monde** [comment]
ok @JPLachance , setting client_id on the provider at the same time as setting the api token is a conflict https://github.com/okta/terraform-provider-okta/blob/master/okta/provider.go#L146-L152

The error message generated by the terraform runtime isn't terribly helpful:

```shell
│ Error: Conflicting configuration arguments
│
│   with provider["registry.terraform.io/okta/okta"],
│   on <empty> line 0:
│   (source code not available)
│
│ "api_token": conflicts with client_id
```

I'm going to update the documentation in a PR so it is clearly stated in the documentation https://registry.terraform.io/providers/okta/okta/latest/docs

I'll close this issue when that gets merged. Let me know if you are good to go.

**Comment by JPLachance** [comment]
Hello @monde,

In my example, I was **not** using the `OKTA_API_TOKEN` environment variable, AKA, I was not setting an API Token and I was only using the Client ID, private key and scopes combo, like described in the provider documentation.

What I did:

- I created an application

<img width="1051" alt="Screen Shot 2022-02-25 at 8 10 22 AM" src="https://user-images.githubusercontent.com/3448802/155720792-ff52b1f7-ff59-4a24-938b-da2b6aa50dc2.png">

- Gave that application credentials to the provider
- The `terraform plan` works
- The `terraform apply` doesn't

If I use the same config you did, AKA if I set a value to OKTA_API_TOKEN and ditch the Client ID, private key and scopes, yes, the provider works.

So, my bug is more around: What is the recommended setup to use the Okta Terraform provider in a CI/CD environment?
- A service account (AKA a dedicated user in our directory) with an API token?
- An application with a Client ID, private key and scopes?

From my initial experiments, it looks like the provider is not fully functional when we use an Application credentials to use it.

Therefore, if I'm right, I would make it clear in the documentation that an Application credentials is not privileged enough to make the provider work. I would document:
- When to use `client_id`, `private_key` and `scopes`?
- Why do they they exist?
- Use-cases?

What do you think?

**Comment by monde** [comment]
@JPLachance you're creating an application with the `okta_app_saml` resource. To do so, you use your organization's token `api_token`, that the is highest level token. As far as `client_id`, `private_key`, those are values that are for an existing application, therefore the they are mutually exclusive to `api_token`. `scopes` are used in the context of an application, creating and updating settings/permissions on the application and its related resources.

In the Okta UI console screen shot of "Create new app integration", when you finish clicking through the settings for that form and get a new application with an id and secret like the `My API Services App` example below. "Client ID" is `client_id` and "Client secret" is `private_key`.

![image](https://user-images.githubusercontent.com/201/155747988-509452c3-8bdb-4f33-b05c-834134603679.png)

For your CI/CD are you needing to create applications or work with existing applications for your pipeline, maybe both? You'd have to split up your Terraforms that interact with existing applications and set the provider with `client_id`, `private_key` and `scopes` for the specific application for those tests. Only use `api_token` when you are testing modifying your our and creating applications.

I'll add these details to the documentation PR #987 I started.

**Comment by JPLachance** [comment]
Hello,

Ok. So then I tried to follow [this blog post](https://matthewdavis111.com/terraform/okta-terraform/) and then I faced a new issue: https://help.okta.com/en/prod/Content/Topics/Security/administrators-admin-comparison.htm

An "App Admin" is not allowed to create an API Token 😅 

It feels like creating a SAML application using the Okta Terraform provider can only be done by a "Super Admin". Am I reading this right?

My goal is simple:
- Let developers register their own apps without granting them access to manage access rights

It feels like I cannot achieve this goal today.

Am I right?

**Comment by monde** [comment]
@JPLachance yes, when I was demo'ing my tf scripts my account is a super user on my test org, so my api token is that of the super user. When I look at the "Application management" section of https://help.okta.com/en/prod/Content/Topics/Security/administrators-admin-comparison.htm it seems to me a user can be granted app management permissions. So an api token created for that user should be able to create applications.

So it seems to me, you'd need to have your super admin generate api tokens for the users with application permissions so they can use that token.

On your requirements "Let developers register their own apps without granting them access to manage access rights", are those developers in your own company? Are you a super admin on your Okta org? Do the apps already exist, register implies they already exist, no?

**Comment by JPLachance** [comment]
Hello,

With API tokens creation, permissions apply to self only. Therefore, a super user cannot create an API token for an application admin user.

I'm a super admin. My goal is to delegate application creation to the application owner. For example, our Infra team owns an OpenSearch cluster. Therefore, I want to give them the ability, through Terraform, to create their own apps in Okta for their OpenSearch clusters.

So today in my orgs, I don't have any of their OpenSearch cluster. I want them to use Terraform to create those, then I'll use Terraform to assign permissions.

Thanks for your help!

**Comment by monde** [comment]
I'm looking for some additional experience internally at Okta with your requirements @JPLachance , I will report back what I find. Thanks

**Comment by JPLachance** [comment]
Hello!

I'm sorry, but I do not really understand how #987 fixed my issues 😅 

Any updates on you internal discussions?

**Comment by noinarisak** [comment]
hey @JPLachance !

Before we start digging to your end goal. Were you able to get the `private_key` option working?

**Comment by monde** [comment]
Glad to have @noinarisak sharing some additional experience. Just want to be frank and point out github issues are for bugs in our open source code as well as asks for new features in these SDK related code paths. And that it seems to me the problem that is trying to find a resolution here is related to the capabilities of the Okta management API and user/role permissions models therein.

We did do an Okta terraform provider release yesterday and I refined the explanation of `api_token` vs `client_id`, `private_key`, and `scopes` from PR #987 that closed this issue and is included in the release.

**Comment by JPLachance** [comment]
Hello,

An Application with a Private Key and Client ID fails to "create" a new application. When I tried, I got the "invalid private key" error.

That being said, when we look at the tooltip in the UI, it suggests it can work.
<img width="1061" alt="Screen Shot 2022-03-04 at 12 36 31 PM" src="https://user-images.githubusercontent.com/3448802/156813586-097bd4b9-5f8c-4807-b722-a94fb5ae96c3.png">

Therefore, I don't know where the real error is. Is the `okta.apps.manage` Okta API scope allowed to create an application, or not? If it is, then why is the provider telling me "invalid private key"?

Thanks!

**Comment by monde** [comment]
@JPLachance if you run terraform with `TF_LOG=1` you can see an actual representation of the HTTP request and response on the command e.g. `TF_LOG=1 terraform apply`. This is helpful for diagnosing support questions.

**Comment by JPLachance** [comment]
Hello,

I found the problem when I read https://github.com/okta/terraform-provider-okta/issues/892#issuecomment-1009133141.

Basically, I was using a client secret in the private_key field, hence why the "invalid private key" message.

Like @noinarisak wrote in https://github.com/okta/terraform-provider-okta/issues/892#issuecomment-1036387631, better documentation is required.

Anyways, case closed.
